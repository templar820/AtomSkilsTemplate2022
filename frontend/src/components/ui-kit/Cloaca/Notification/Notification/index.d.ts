import * as React from 'react';
import { EmotionStylesType } from 'lib-ui';
import { MouseEvent } from 'react';

export interface NotificationProps {
    /** Defines preset of component */
    preset?: 'success' | 'attention' | 'error' | 'custom';
    /** Defines color of component */
    notificationColor?: string;
    /** Defines icon of component */
    icon?: React.ReactNode;
    /** Defines text or any node that will be placed in a body of the component */
    children?: React.ReactNode;
    /** Defines whether component should be filled */
    isFilled?: boolean;
    /** Callback that will be fired on clicking notification's body
     * If notification is rendered by NotificationContainer, gets called with second argument -
     * callback to close notification if necessary
     * */
    onClick?: (event: MouseEvent, closeNotification: () => void) => void;
    /** Callback that will be fired on clicking notification's close icon */
    /** If returns `true` - notification won't be closed */
    onClose?: (event: MouseEvent) => void | true;
    /** Defines font-size */
    size?: string | number;
    /** Defines additional className */
    className?: string;
    /** Defines additional component css */
    css?: EmotionStylesType;
    /** time to live in the hidden state (ms) */
    ttl?: number;
}

declare const Notification: React.ComponentType<NotificationProps>;

export default Notification;
