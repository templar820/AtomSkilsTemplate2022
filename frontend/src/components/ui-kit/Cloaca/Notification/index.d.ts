export { default as Notification } from './Notification';
export * from './Notification';

export { default as NotificationContainer } from './NotificationContainer';
export * from './NotificationContainer';

export { default as NotificationManager } from './NotificationManager';
export * from './NotificationManager';
