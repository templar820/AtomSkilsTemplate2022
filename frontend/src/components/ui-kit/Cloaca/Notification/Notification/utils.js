const configs = {
    success: {
        notificationColor: 'primary',
        children: 'Success',
        icon: 'Success'
    },
    attention: {
        notificationColor: 'warning',
        children: 'Attention',
        icon: 'Attention_error',
        iconForFilledState: 'Attention_outlined'
    },
    error: {
        notificationColor: 'error',
        children: 'Error',
        icon: 'Attention_error',
        iconForFilledState: 'Attention_outlined'
    },
    custom: {}
};

function getNotificationConfig(inputConfig, colors) {
    const { preset, ...restInputConfig } = inputConfig;
    if (!configs[preset]) console.warn(`There are no such notification preset as "${preset}".`);
    const { notificationColor, ...restPresetConfig } = configs[preset];
    const _notificationColor = colors[notificationColor] || notificationColor;

    return { notificationColor: _notificationColor, ...restPresetConfig, ...restInputConfig };
}

export { getNotificationConfig };
