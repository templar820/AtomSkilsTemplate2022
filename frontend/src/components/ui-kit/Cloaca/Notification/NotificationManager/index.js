import uuidv4 from 'uuid/v4';

class NotificationManager {
    notificationList = [];
    notificationWaitingList = [];
    changeListeners = {};
    autoDispatchChanges = true;
    defaultDuration = 5000;
    defaultNotification = { preset: 'attention' };
    maxShowed = 3;

    setDefaultDuration = (duration) => {
        this.defaultDuration = duration;
    };

    setDefaultNotification = (notification) => {
        this.defaultNotification = notification;
    };

    setAutoDispatchChanges = (boolean) => {
        this.autoDispatchChanges = boolean;
    };

    setMaxShowed = (num) => {
        this.maxShowed = num;
    };

    clearTimeoutNotifications = (notificationList) => {
        notificationList.forEach((notification) => {
            clearTimeout(notification.timeoutId);
        });
    };

    // Clears all lists of notifications
    clearNotificationList = () => {
        this.clearTimeoutNotifications(this.notificationList);

        this.notificationList = [];
        this.notificationWaitingList = [];

        if (this.autoDispatchChanges) this.dispatchChanges();
    };

    clearChangeListeners = () => {
        this.changeListeners = {};
    };

    /** Adds notification to the notificationWaitingList. */
    _addWaitingNotification = ({ ttl, ...notification }) => {
        // If notification has ttl option sets timer to remove the notification.
        if (ttl) {
            const timeoutId = setTimeout(() => {
                this._removeWaitingNotificationById(notification.id);
            }, ttl);

            this.notificationWaitingList.push({ ...notification, timeoutId });
            return;
        }

        this.notificationWaitingList.push(notification);
    };

    /** Removes notification from the notificationWaitingList. */
    _removeWaitingNotificationById = (id) => {
        const index = this.notificationWaitingList.findIndex((item) => item.id === id);

        this.notificationWaitingList.splice(index, 1);
    };

    /** Adds notification to the notificationList.
     *  If the list is full, add to the notificationWaitingList. */
    addNotification = (notification) => {
        const { duration = this.defaultDuration, ...restNotification } = notification;
        const newNotification = {
            id: uuidv4(),
            ...this.defaultNotification,
            ...restNotification
        };

        // Add notification to the waiting list if notificationList is full
        if (this.notificationList.length >= this.maxShowed) {
            this._addWaitingNotification(newNotification);
            return;
        }

        // Clear ttl-timer for notification from the waiting list with ttl option
        if (notification.timeoutId) {
            clearTimeout(notification.timeoutId);
        }

        this.notificationList.push(newNotification);

        if (duration !== 0) {
            const timeoutId = setTimeout(() => {
                this.removeNotificationById(newNotification.id);
            }, duration);

            newNotification.timeoutId = timeoutId;
        }

        if (this.autoDispatchChanges) this.dispatchChanges();
    };

    /** Removes notification from the notificationList */
    removeNotificationById = (id) => {
        const index = this.notificationList.findIndex((item) => {
            const condition = item.id === id;
            condition && clearTimeout(item.timeoutId);
            return condition;
        });

        this.notificationList.splice(index, 1);

        // Add new notification to the notificationList from notificationWaitingList
        if (this.notificationWaitingList.length) {
            this.addNotification(this.notificationWaitingList[0]);
            this.notificationWaitingList.splice(0, 1);
        }

        if (this.autoDispatchChanges) this.dispatchChanges();
    };

    dispatchChanges = () => {
        Object.keys(this.changeListeners).forEach((callback) => {
            this.changeListeners[callback](this.notificationList);
        });
    };

    addChangeListener = (callback) => {
        this.changeListeners[callback] = callback;
    };

    removeChangeListener = (callback) => {
        delete this.changeListeners[callback];
    };
}

export default new NotificationManager();
export { NotificationManager };
