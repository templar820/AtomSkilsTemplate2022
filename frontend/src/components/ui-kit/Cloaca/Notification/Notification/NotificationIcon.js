/** @jsx jsx */
import { jsx } from '@emotion/core';
import { px } from 'lib-root/utils/units';
import React from 'react';
import InlineIcons from 'lib-ui/InlineIcons';

const NotificationIcon = React.forwardRef(
    ({ size, icon, notificationColor, isFilled, iconForFilledState, clickable, onClick }, ref) => {
        const styles = {
            marginRight: icon !== 'Cross_exit' ? px(size, 1.0667) : 0,
            marginLeft: icon === 'Cross_exit' ? 'auto' : 0,
            marginTop: icon === 'Success' ? '-2px' : '1px',
            cursor: clickable ? 'pointer' : 'default'
        };
        const iconSize =
            icon === 'Success'
                ? px(size, 1.6)
                : icon === 'Cross_exit'
                ? px(size, 0.9333333333333333)
                : px(size, 1.3333333333333333);
        return (
            <InlineIcons
                ref={ref}
                onClick={onClick}
                icon={isFilled ? iconForFilledState || icon : icon}
                color={isFilled ? 'currentColor' : notificationColor}
                style={styles}
                w={iconSize}
                h={iconSize}
            />
        );
    }
);

export default NotificationIcon;
