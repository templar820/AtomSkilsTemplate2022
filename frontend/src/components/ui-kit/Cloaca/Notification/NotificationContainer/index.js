import React, { useReducer, useEffect } from 'react';
import { Transition } from 'react-transition-group';
import PropTypes from 'prop-types';

import Portal from 'lib-ui/utils/Portal/Portal';

import { setClassName } from 'lib-root/utils/styleMixins';
import { canUseDOM } from 'lib-root/utils';
import { LayersContext } from 'lib-root/layers';

import Notification from '../Notification';
import NotificationManager from '../NotificationManager';
import NotificationPositioner from './NotificationPositioner';
import { StyledNotificationContainer } from './units';

const notificationsReducer = (_, payload) => [...payload];

const NotificationContainer = React.forwardRef(
    ({ className, css: _css, horizontalPosition, x, verticalPosition, y, ...restProps }, ref) => {
        const [notifications, dispatchNotifications] = useReducer(notificationsReducer, []);

        useEffect(() => {
            NotificationManager.addChangeListener(dispatchNotifications);
            return () => {
                NotificationManager.removeChangeListener(dispatchNotifications);
            };
        });

        const renderedNotifications = notifications.map(({ id, onClick, onClose, ...restNotificationProps }) => (
            <Notification
                key={id}
                {...restNotificationProps}
                onClose={(e) => {
                    let preventClosing;
                    if (typeof onClose === 'function') preventClosing = onClose(e);
                    if (!preventClosing) NotificationManager.removeNotificationById(id);
                }}
                onClick={(e) => {
                    if (onClick) onClick(e, () => NotificationManager.removeNotificationById(id));
                }}
            />
        ));

        return (
            <Portal node={document.body} roleName={'notification'} hasLayerControl={false}>
                <LayersContext.Consumer>
                    {(zIndex) => (
                        <StyledNotificationContainer
                            {...{ ref, _css, zIndex }}
                            className={setClassName({ props: { className }, name: 'notification__container' })}>
                            <Transition in={!!notifications.length && canUseDOM} timeout={500}>
                                {(state) => (
                                    <NotificationPositioner
                                        horizontalPosition={x || horizontalPosition}
                                        verticalPosition={y || verticalPosition}
                                        notificationsLength={notifications.length}
                                        {...{ state, ...restProps }}>
                                        {renderedNotifications}
                                    </NotificationPositioner>
                                )}
                            </Transition>
                        </StyledNotificationContainer>
                    )}
                </LayersContext.Consumer>
            </Portal>
        );
    }
);

NotificationContainer.displayName = 'NotificationContainer';
NotificationContainer.propTypes = {
    /** Defines horizontal position */
    horizontalPosition: PropTypes.oneOf(['left', 'right']),
    /** Defines vertical position */
    verticalPosition: PropTypes.oneOf(['top', 'bottom']),
    /** alias for horizontalPosition */
    x: PropTypes.oneOf(['left', 'right']),
    /** alias for verticalPosition */
    y: PropTypes.oneOf(['top', 'bottom']),
    /** Defines direction of component */
    direction: PropTypes.oneOf(['column', 'row']),
    /** Defines whether component should be reversed */
    isReversed: PropTypes.bool,
    /** Defines padding of component */
    gap: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    /** Defines additional className */
    className: PropTypes.string,
    /** Defines additional component css */
    css: PropTypes.oneOfType([PropTypes.func, PropTypes.object, PropTypes.string]),
    /** Defines additional emotion styles for close icon */
    customStylesForCloseButton: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
    /** Defines whether closeAll should be shown */
    closeAll: PropTypes.oneOfType([PropTypes.oneOf([null]), PropTypes.node, PropTypes.string]),
    /** Defines number of notifications needed for closeAll to be shown */
    countForView: PropTypes.number
};
NotificationContainer.defaultProps = {
    horizontalPosition: 'right',
    verticalPosition: 'bottom',
    direction: 'column',
    isReversed: true,
    gap: '12',
    closeAll: 'Закрыть все оповещения',
    countForView: 1
};

export default NotificationContainer;
