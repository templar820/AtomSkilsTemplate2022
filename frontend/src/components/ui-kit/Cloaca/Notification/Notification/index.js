import React, { useRef } from 'react';
import PropTypes from 'prop-types';

import styled from '@emotion/styled';

import withColors from 'lib-ui/utils/withColors';
import { autoAddPx } from 'lib-root/utils/styleMixins';

import NotificationIcon from './NotificationIcon';
import { getNotificationConfig } from './utils';
import { setClassName } from 'lib-root/utils/styleMixins';
import applyEmotionStyle from 'lib-root/utils/applyEmotionStyle';

const StyledNotification = styled.div`
    display: inline-flex;
    min-width: 456px;
    max-width: 688px;
    padding: ${({ isFilled }) => (isFilled ? `3px 16px` : `14px 16px 8px`)};
    border: 1px solid ${({ notificationColor }) => notificationColor};
    box-shadow: 0 0 8px rgba(51, 51, 51, 0.25);
    border-left-width: 8px;
    background-color: ${({ colors, isFilled, notificationColor }) =>
        isFilled ? notificationColor : colors.GrayScale_0};
    color: ${({ colors, isFilled }) => (isFilled ? colors.GrayScale_0 : 'currentColor')};
    font-size: ${({ size }) => autoAddPx(size)};
    line-height: 1.6;
    pointer-events: auto;
    ${({ css, ...restProps }) => applyEmotionStyle(css, restProps)};
`;

/**
 * Компонент всплывающего окна/уведомления.
 *
 * import { Notification as NotificationComponents } from 'core-lib-react/components';
 *
 * const { Notification, NotificationContainer, NotificationManager } = NotificationComponents;
 */
const Notification = React.forwardRef(({ colors, size, ...restProps }, ref) => {
    const { className, children, onClick, onClose, ...restNotificationConfig } = getNotificationConfig(
        restProps,
        colors
    );
    const closeRef = useRef();
    return (
        <StyledNotification
            onClick={(e) => {
                if (e.target !== closeRef.current && !closeRef.current.contains(e.target)) {
                    onClick && onClick(e);
                }
            }}
            {...{ ref, colors, size, ...restNotificationConfig }}
            className={setClassName({ props: { className }, name: 'notification' })}>
            <NotificationIcon {...{ size, ...restNotificationConfig }} />
            {children}
            <NotificationIcon
                {...{
                    size,
                    ...restNotificationConfig,
                    clickable: true,
                    ref: closeRef,
                    onClick: (e) => onClose && onClose(e),
                    icon: 'Cross_exit',
                    iconForFilledState: 'Cross_exit'
                }}
            />
        </StyledNotification>
    );
});

Notification.displayName = 'Notification';
Notification.propTypes = {
    /** Defines preset of component */
    preset: PropTypes.oneOf(['success', 'attention', 'error', 'custom']),
    /** Defines color of component */
    notificationColor: PropTypes.string,
    /** Defines icon of component */
    icon: PropTypes.oneOfType([PropTypes.node, PropTypes.string]),
    /** Defines text or any node that will be placed in a body of the component */
    children: PropTypes.node,
    /** Defines whether component should be filled */
    isFilled: PropTypes.bool,
    /** Callback that will be fired on clicking notification's body */
    onClick: PropTypes.func,
    /** Callback that will be fired on clicking notification's close icon */
    onClose: PropTypes.func,
    /** Defines font-size */
    size: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    /** Defines additional className */
    className: PropTypes.string,
    /** Defines additional component css */
    css: PropTypes.oneOfType([PropTypes.func, PropTypes.object, PropTypes.string]),
    /** time to live in the hidden state (ms) */
    ttl: PropTypes.number
};
Notification.defaultProps = {
    preset: 'attention',
    size: '15px',
    isFilled: true
};

export default withColors(Notification);
