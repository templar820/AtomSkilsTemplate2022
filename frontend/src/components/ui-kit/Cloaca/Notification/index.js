import Notification from './Notification';
import NotificationContainer from './NotificationContainer';
import NotificationManager from './NotificationManager';

export { Notification, NotificationContainer, NotificationManager };
