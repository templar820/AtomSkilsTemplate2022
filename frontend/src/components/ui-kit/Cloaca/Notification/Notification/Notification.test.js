import React from 'react';
import ReactDOM from 'react-dom';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';

import Notification from './';
import { colors } from 'lib-root/colors';

Enzyme.configure({ adapter: new Adapter() });

const mockFn = jest.fn();

const props = {
    preset: 'error',
    notificationColor: colors.errorAccent,
    icon: 'Archive',
    children: <div id="test7">Notification test7 content</div>,
    isFilled: false,
    size: 17,
    className: 'some-class',
    css: { fontSize: '18px' },
    ttl: 3001,
    onClick: mockFn
};

describe('<Notification />', () => {
    let wrapper;

    beforeEach(() => {
        wrapper = shallow(<Notification {...props} />);
    });

    afterEach(() => {
        wrapper.unmount();
    });

    // test must be first, otherwise the snapshot size will be too large
    test('should match snapshot', () => {
        expect(toJson(wrapper.dive())).toMatchSnapshot();
    });

    test('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(<Notification {...props} />, div);
        ReactDOM.unmountComponentAtNode(div);
    });

    test('should render correctly', () => {
        expect(wrapper.find('NotificationIcon')).toBeTruthy();
        expect(wrapper.find('div[id="test7"]')).toBeTruthy();
        expect(wrapper.find('Notification test7 content')).toBeTruthy();
        expect(wrapper.find('.some-class')).toBeTruthy();

        expect(wrapper.dive().props().onClick).toBeDefined();
        expect(typeof wrapper.dive().props().onClick).toBe('function');
        expect(wrapper.dive().props()).toHaveProperty('isFilled', props.isFilled);
        expect(wrapper.dive().props()).toHaveProperty('notificationColor', props.notificationColor);
        expect(wrapper.dive().props()).toHaveProperty('icon', props.icon);
        expect(wrapper.dive().props()).toHaveProperty('size', props.size);
        expect(wrapper.dive().props()).toHaveProperty('ttl', props.ttl);
        expect(wrapper.dive().props()).toHaveProperty('css', props.css);
    });

    test('should call onClick method', () => {
        wrapper.dive().simulate('click');
        expect(mockFn).toHaveBeenCalled();
    });
});
