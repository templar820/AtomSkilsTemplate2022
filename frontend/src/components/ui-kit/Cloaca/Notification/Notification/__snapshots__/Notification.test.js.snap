// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Notification /> should match snapshot 1`] = `
<Notification
  className="some-class"
  colors={
    Object {
      "GrayScale_0": "#ffffff",
      "GrayScale_100": "#EDEDED",
      "GrayScale_150": "#E5E5E5",
      "GrayScale_200": "#D8D8D8",
      "GrayScale_50": "#fafafa",
      "GrayScale_500": "#999999",
      "GrayScale_55": "#f6f6f6",
      "GrayScale_600": "#7C7C7C",
      "GrayScale_700": "#333333",
      "GrayScale_900": "#000000",
      "error": "#F62434",
      "errorAccent": "#EB5A40",
      "info": "#34AAF2",
      "infoAccent": "#5BD9E5",
      "primary": "#00B956",
      "primaryAccent": "#10E272",
      "secondary": "#731982",
      "secondaryAccent": "#A500BF",
      "tertiary": "#444189",
      "tertiaryAccent": "#404D46",
      "warning": "#FFA717",
      "warningAccent": "#F36F20",
    }
  }
  css={
    Object {
      "fontSize": "18px",
    }
  }
  icon="Archive"
  isFilled={false}
  notificationColor="#EB5A40"
  onClick={[MockFunction]}
  preset="error"
  size={17}
  ttl={3001}
>
  <div
    id="test7"
  >
    Notification test7 content
  </div>
</Notification>
`;
